### PR TITLE
feat(web,renderer-app,flat-component): add agora sso

### PR DESCRIPTION
**Pull request by Matrixbirds**
1. add Agora SSO Button.
2. integrate SSO FlatServer Backend API.
3. add `utm_source` parameter switch with Agora SSO Button. 
4. refactor LoginPanel & LoginContent components.

### CODE REVIEW

File: desktop/renderer-app/src/pages/LoginPage/index.tsx [hunk 1]
```diff
@@ -76,20 +75,67 @@ export const LoginPage = observer(function LoginPage() {
...
+                loginDisposer.current = void 0;
+            }
+            if (agreementChecked) {
+                doLogin(loginChannel);
+            } else {
+                void message.info(i18n.t("agree-terms"));
+            }
+        },
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        [agreementChecked],
```
Comment (BlackHole1) [inline desktop/renderer-app/src/pages/LoginPage/index.tsx]:
```suggestion
        [agreementChecked, message],
```

File: packages/flat-components/src/components/LoginPage/LoginContent/LoginContent.stories.tsx [hunk 2]
```diff
@@ -9,8 +12,44 @@ const storyMeta: Meta = {
...
+                <LoginButton provider={"github"} text={t("login-github")} onLogin={handleLogin} />
+            </>
+        );
+    }
+
+    return (
+        <div className="vh-75">
+            <LoginContent
+                agreementChecked={agreement}
+                handleClickAgreement={handleClickAgreement}
```
Comment (BlackHole1) [inline packages/flat-components/src/components/LoginPage/LoginContent/LoginContent.stories.tsx]:
```suggestion
                handleClickAgreement={() => updateAgreement(!agreement)}
```

File: packages/flat-components/src/components/LoginPage/LoginContent/LoginContent.stories.tsx [hunk 3]
```diff
@@ -9,8 +12,44 @@ const storyMeta: Meta = {
...
+        updateAgreement(!agreement);
+    };
+
+    const handleLogin = useCallback(() => {
+        agreement === false && void message.info(i18n.t("agree-terms"));
+    }, [agreement, i18n]);
+
+    const handleHideQRCode = (): void => {
+        updateShowQRCode(!showQRCode);
+    };
```
Comment (BlackHole1) [inline packages/flat-components/src/components/LoginPage/LoginContent/LoginContent.stories.tsx]:
```suggestion
```

File: packages/flat-components/src/components/LoginPage/LoginPanel.stories.tsx [hunk 4]
```diff
@@ -31,26 +30,63 @@ const storyMeta: Meta = {
...
+                        Modal.info({ content: "This is Github Login" });
+                        return;
+                    }
+                    default: {
+                        return;
+                    }
+                }
+            } else {
+                void message.info(i18n.t("agree-terms"));
+            }
```
Comment (BlackHole1) [inline packages/flat-components/src/components/LoginPage/LoginPanel.stories.tsx]:
```suggestion
            if (!agreement) {
                void message.info(i18n.t("agree-terms"));
                return;
            }
            
            switch (loginChannel) {
                case "wechat": {
                    setWeChatLogin(true);
                    break;
                }
                case "github": {
                    Modal.info("need i18n");
                    break;
                }
                default: {
                    break;
                }
            }
```

File: web/flat-web/src/pages/LoginPage/index.tsx [hunk 5]
```diff
@@ -29,13 +36,23 @@ export const LoginPage = observer(function LoginPage() {
...
+                    globalStore.updateUserInfo(authData);
+                    if (roomUUID) {
+                        if (globalStore.isTurnOffDeviceTest) {
+                            await joinRoomHandler(roomUUID, pushHistory);
+                        } else {
+                            pushHistory(RouteNameType.DevicesTestPage, { roomUUID });
+                        }
+                    } else {
+                        pushHistory(RouteNameType.HomePage);
+                    }
```
Comment (BlackHole1) [inline web/flat-web/src/pages/LoginPage/index.tsx]:
```suggestion
                    if (!roomUUID) {
                        pushHistory(RouteNameType.HomePage);
                        return;
                    }
                    
                    if (globalStore.isTurnOffDeviceTest) {
                        await joinRoomHandler(roomUUID, pushHistory);
                    } else {
                        pushHistory(RouteNameType.DevicesTestPage, {
                            roomUUID
                        });
                    }
```

File: web/flat-web/src/pages/LoginPage/index.tsx [hunk 6]
```diff
@@ -52,20 +69,89 @@ export const LoginPage = observer(function LoginPage() {
...
+                loginDisposer.current = void 0;
+            }
+            if (agreement) {
+                doLogin(loginChannel);
+            } else {
+                void message.info(i18n.t("agree-terms"));
+            }
+        },
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        [agreement],
```
Comment (BlackHole1) [inline web/flat-web/src/pages/LoginPage/index.tsx]:
```suggestion
        [agreement, message],
```

File: web/flat-web/src/pages/LoginPage/index.tsx [hunk 7]
```diff
@@ -52,20 +69,89 @@ export const LoginPage = observer(function LoginPage() {
...
+                    <LoginButton
+                        provider="wechat"
+                        text={i18n.t("login-wechat")}
+                        onLogin={handleLogin}
+                    />
+                    <LoginButton
+                        provider="github"
+                        text={i18n.t("login-github")}
+                        onLogin={handleLogin}
+                    />
```
Comment (BlackHole1) [inline web/flat-web/src/pages/LoginPage/index.tsx]:
```suggestion
```

File: web/flat-web/src/utils/routes.ts [hunk 8]
```diff
@@ -77,3 +77,20 @@ export function useReplaceHistory(): <T extends RouteNameType>(
 
     return pushHistory;
 }
+
+/**
+ * get url parameters
+ */
+export function useURLParams(): Record<string, any> {
```
Comment (BlackHole1) [inline web/flat-web/src/utils/routes.ts]: react-router hook?

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: packages/flat-components/src/components/LoginPage/index.less [hunk 9]
```diff
@@ -46,4 +46,4 @@
 
 .login-panel-inner-content-container {
     margin-top: 178px;
-}
+}
```
Comment (Cheerego7) [inline packages/flat-components/src/components/LoginPage/index.less]: add newline

File: packages/flat-components/src/components/LoginPage/LoginButton/index.less [hunk 10]
```diff
@@ -44,4 +63,4 @@
             background: #1b1f22;
         }
     }
-}
+}
```
Comment (Cheerego7) [inline packages/flat-components/src/components/LoginPage/LoginButton/index.less]: add newline

**Review by Cheerego7 — DISMISSED** [review]

**Comment by Matrixbirds** [comment]
cc @BlackHole1

**Comment by Matrixbirds** [comment]
@BlackHole1  any progress?

**Review by BlackHole1 — APPROVED** [review]

**Pull request merged**